x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -158,9 +165,24 @@ public void onFailure(Exception e) {
 
             @Override
             protected void doRun() {
-                restoreService.restoreSnapshot(restoreRequest,
-                    ActionListener.delegateFailure(listener,
-                        (delegatedListener, response) -> afterRestoreStarted(clientWithHeaders, request, delegatedListener, response)));
+                ActionListener<RestoreService.RestoreCompletionResponse> delegatelistener = ActionListener.delegateFailure(
+                    listener,
+                    (delegatedListener, response) -> afterRestoreStarted(clientWithHeaders, request, delegatedListener, response)
+                );
+                if (remoteDataStream == null) {
+                    restoreService.restoreSnapshot(restoreRequest, delegatelistener);
+                } else {
+                    String followerIndexName = request.getFollowerIndex();
+                    BiConsumer<ClusterState, Metadata.Builder> updater = (currentState, mdBuilder) -> {
```

Comment:
Can we make this change directly in `RestoreService`? A SnapshotInfo from CcrReposity should contain the data streams from the leader cluster?

Reply:
Let me see if that works. If so, then I agree that is a cleaner change.

Reply:
The `RestoreService` currently only restores the data stream metadata if a data stream needs to be restored or an entire snapshot. In the case that a specific backing index is restored the associated data stream isn't restored. The put follow only 'restores' a backing index at a time and if this logic would be included in `RestoreService` then I think we always need to load and check global metadata state in order to determine whether an index is a backing index of a data stream (when #61525 is merged then some backing indices may not follow the backing index naming scheme).

I think we would need to introduce additional parameter on `RestoreSnapshotRequest` to indicate that we intent to restore a backing index and therefor the associated data stream should be restored as well (or existing local data stream should be updated). Otherwise regular restore operations would always need to load the global metadata state in order to determine whether an index is a backing index of a data stream. Do you think that this is still an improvement over the current logic?

Reply:
Ok, let's keep the current implementation. We can look into the improvement once the datastream becomes mature.